src/main/scala/com/github/codelionx/dodo/discovery/IndexedOrdering.scala

```diff
@@ -0,0 +1,98 @@
+package com.github.codelionx.dodo.discovery
+
+import com.github.codelionx.dodo.types.{DataType, TypedColumn}
+
+import scala.collection.mutable
+
+
+trait IndexedOrdering {
+
+  /**
+    * Returns the original indices of the dataset in the order when sorting by the supplied columns lexicographically.
+    */
+  def orderedIndicesOf(dataset: Array[TypedColumn[_ <: Any]], sortingColumns: Seq[Int]): Seq[Int] = {
+    if (sortingColumns.length <= 0 || dataset.length <= 0)
+      Seq.empty
+    else if (sortingColumns.length == 1)
+      dataset(sortingColumns.head).sortedIndices
+    else if (sortingColumns.max >= dataset.length)
+      throw new IllegalArgumentException(
+        "Can not sort by column indices that are greater than the number of columns in the dataset"
+      )
+    else {
+      val bufferView = mutable.Buffer.empty[TypedColumn[_ <: Any]]
+      for (column <- sortingColumns) {
+        bufferView += dataset(column)
+      }
+      recSorting(bufferView.toArray, dataset.head.indices, 0)
```

Comment:
shouldn't the sortingColumns been used here instead of dataset.head.indices?

Reply:
`sortingColumns` are indices to the columns, so if we have a dataset with 4 columns and 20 rows, we can specify `sortingColumns = Seq(3,0)` to sort first for the fourth and then the first column.

`sortIndices` should contain the row indices to be sorted instead. In the example it would be the range `0 until 20`. We could initialize this range here differently:
- by `0 until dataset.head.length`
- by `dataset.head.indices`
- by using a selected column: `dataset(sortingColumns(0)).indices`
Those options would give the same index range.

For deeper (recursive calls) the `sortIndices` value contains a potential non-continuous list of the original row indices of the constant value slice.

But I forgot a test case for this. You are right.